### PR TITLE
Remote upgrades: upgrade agents to the manager version by default

### PR DESCRIPTION
**Pull request by Leamsy**
Edited agent.py so that the agent's maximum version when upgrading by default is always the manager's, allowing the option to select a different version if so desired.

### CODE REVIEW

**Comment by vikman90** [comment]
Related issue: #2458

File: framework/wazuh/agent.py [hunk 1]
```diff
@@ -1966,16 +1966,22 @@ def _get_wpk_file(self, wpk_repo=common.wpk_repo_url, debug=False, version=None,
 
         # Comparing versions
         agent_ver = self.version
-        if debug:
-            print("Agent version: {0}".format(agent_ver.split(" ")[1]))
-            print("Agent new version: {0}".format(agent_new_ver))
+
 
         if WazuhVersion(manager_ver.split(" ")[1]) < WazuhVersion(agent_new_ver):
-            raise WazuhException(1717, "Manager: {0} / Agent: {1} -> {2}".format(manager_ver.split(" ")[1], agent_ver.split(" ")[1], agent_new_ver))
```
Comment (chemamartinez) [inline framework/wazuh/agent.py]: This exception cannot be removed. The agent should not be upgraded to a higher version than the manager without including the force flag `-F`.

**Comment by chemamartinez** [comment]
Found `list index out of range` error when specifying an unexistent version for the agent.

```
# agent_upgrade -a 001 -v v3.9.2 -d
Internal error: list index out of range
Traceback (most recent call last):
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 170, in <module>
    main()
  File "/var/ossec/framework/scripts/agent_upgrade.py", line 124, in main
    rl_timeout=-1 if args.timeout == None else args.timeout, use_http=use_http)
  File "/var/ossec/framework/python/lib/python3.7/site-packages/wazuh-3.8.2-py3.7.egg/wazuh/agent.py", line 2190, in upgrade
    show_progress=show_progress, chunk_size=chunk_size, rl_timeout=rl_timeout, use_http=use_http)
  File "/var/ossec/framework/python/lib/python3.7/site-packages/wazuh-3.8.2-py3.7.egg/wazuh/agent.py", line 2056, in _send_wpk_file
    _get_wpk = self._get_wpk_file(wpk_repo=wpk_repo, debug=debug, version=version, force=force, use_http=use_http)
  File "/var/ossec/framework/python/lib/python3.7/site-packages/wazuh-3.8.2-py3.7.egg/wazuh/agent.py", line 1953, in _get_wpk_file
    if versions[0] == version:
IndexError: list index out of range
```

This bug is shared with the stable version.

**Comment by chemamartinez** [comment]
Last two commits add the following fixes:

[7f5fb12e6f6f1753c275b1eb07afafd9f6ffdbdd] -> Looks for the manager version first of all if the version to upgrade to is not specified.
[d5d9b195e275bf6d7622c40c254cf59c05766b6b] -> Fixed the error reported at https://github.com/wazuh/wazuh/pull/2721#issuecomment-492572300 due to the list of available versions contained an empty item:

```
[['v3.9.0', '180e25a1fefafe8d83c763d375cb1a3a387bc08a'], ['v3.8.2', '7a49d5604e1034d1327c993412433d124274bc7e'], ['v3.8.1', '54c55d50f9d88df937fb2b40a4eeec17cbc6ce24'], ['v3.7.2', 'e28cfb89469b1b8bfabefe714c09b942ebd7a928'], ['v3.7.1', '7ef661a92295a02755812e3e10c87bf49bb52114'], ['v3.7.0', 'b1a94c212195899be53564e86b69981d4729154e'], ['v3.6.1', 'ed01192281797f64c99d53cff91efe936bc31b17'], ['v3.6.0', '83fd0e49c6ab47f59c5d75478a371396082613fe'], ['v3.5.0', '5e276bd26d76c3c1eebed5ca57094ee957b3ee40'], ['v3.4.0', 'f20e4319b9088d534a4655a9136a608800522d50'], []]
```

### Performed test

Tested with an agent v3.7.0 and a manager v3.8.2

- [x] **3.7.0 -> manager version**

```
# agent_upgrade -a 001 -d
Manager version: v3.8.2
Agent version: v3.7.0
Agent new version: v3.8.2
WPK file already downloaded: /var/ossec/var/upgrade/wazuh_agent_v3.8.2_linux_x86_64.wpk - SHA1SUM: 7a49d5604e1034d1327c993412433d124274bc7e
Upgrade PKG: wazuh_agent_v3.8.2_linux_x86_64.wpk (17008.7822265625 KB)
MSG SENT: 001 com open wb wazuh_agent_v3.8.2_linux_x86_64.wpk
RESPONSE: ok
MSG SENT: 001 com lock_restart -1
RESPONSE: ok
Chunk size: 512 bytes
Sending: /var/ossec/var/upgrade/wazuh_agent_v3.8.2_linux_x86_64.wpk
Sending WPK: [=                        ] 4%
```

- [x] **3.7.0 -> Latest version without force option**

```
# agent_upgrade -a 001 -v v3.9.0
Error 1717: Cannot upgrade to a version higher than the manager: Manager: v3.8.2 / Agent: v3.7.0 -> v3.9.0
```

- [x] **3.7.0 -> Latest version with force option**

```
# agent_upgrade -a 001 -v v3.9.0 -F -d
Manager version: v3.8.2
Agent version: v3.7.0
Agent new version: v3.9.0
WPK file already downloaded: /var/ossec/var/upgrade/wazuh_agent_v3.9.0_linux_x86_64.wpk - SHA1SUM: 180e25a1fefafe8d83c763d375cb1a3a387bc08a
Upgrade PKG: wazuh_agent_v3.9.0_linux_x86_64.wpk (4492.53125 KB)
MSG SENT: 001 com open wb wazuh_agent_v3.9.0_linux_x86_64.wpk
RESPONSE: ok
MSG SENT: 001 com lock_restart -1
RESPONSE: ok
Chunk size: 512 bytes
Sending: /var/ossec/var/upgrade/wazuh_agent_v3.9.0_linux_x86_64.wpk
Sending WPK: [==                       ] 8%
```

- [x] **Downgrade without force option**

```
# agent_upgrade -a 001 -v v3.3.1
Error 1716: Error upgrading agent: Agent ver: v3.7.0 / Agent new ver: v3.3.1
```

- [x] **Downgrade with force option**

```
# agent_upgrade -a 001 -v v3.3.1 -F -d
Manager version: v3.8.2
Agent version: v3.7.0
Agent new version: v3.3.1
Downloading WPK file from: https://packages.wazuh.com/wpk/centos/7/x86_64/wazuh_agent_v3.3.1_centos_7_x86_64.wpk
WPK file downloaded: /var/ossec/var/upgrade/wazuh_agent_v3.3.1_centos_7_x86_64.wpk - SHA1SUM: f7c953650d2281e9c88c1788ecfef401fd504193
Upgrade PKG: wazuh_agent_v3.3.1_centos_7_x86_64.wpk (4218.064453125 KB)
MSG SENT: 001 com open wb wazuh_agent_v3.3.1_centos_7_x86_64.wpk
RESPONSE: ok
MSG SENT: 001 com lock_restart -1
RESPONSE: ok
Chunk size: 512 bytes
Sending: /var/ossec/var/upgrade/wazuh_agent_v3.3.1_centos_7_x86_64.wpk
Sending WPK: [==                       ] 8%
```

File: framework/wazuh/agent.py [hunk 2]
```diff
@@ -1932,6 +1932,7 @@ def _get_versions(self, wpk_repo=common.wpk_repo_url, version=None, use_http=Fal
 
         if result.ok:
             versions = [version.split() for version in result.text.split('\n')]
+            versions = list(filter(None, versions))
```
Comment (vikman90) [inline framework/wazuh/agent.py]: This sentence looks better:
```python
versions = list(filter(lambda x: len(x) > 0, versions))
```

**Comment by chemamartinez** [comment]
New exception messages:

```
# agent_upgrade -a 001 -v v3.3.1
Error 1749: Downgrading an agent requires the force flag. Use -F to force the downgrade: Agent version: v3.7.0 -> v3.3.1
```

```
# agent_upgrade -a 001 -v v3.9.0
Error 1717: Cannot upgrade to a version higher than the manager. Use the force flag (-F) to achieve it: Manager: v3.8.2 / Agent: v3.7.0 -> v3.9.0
```

With the sentence proposed at https://github.com/wazuh/wazuh/pull/2721#discussion_r284270292 it works fine:

```
# agent_upgrade -a 001 -v v4.9.0
Error 1718: Version not available: v4.9.0
# agent_upgrade -a 001 -v v3.11.0
Error 1718: Version not available: v3.11.0
```

Regards.

**Comment by chemamartinez** [comment]
Final outputs:

### Agent_upgrade tool

```
# agent_upgrade -a 001 -v v4.9.0
Error 1718: Version not available: v4.9.0
# agent_upgrade -a 001 -v v3.9.0
Error 1717: Upgrading an agent to a version higher than the manager requires the force flag. Use -F to force the upgrade: Manager: v3.8.2 / Agent: v3.7.0 -> v3.9.0
# agent_upgrade -a 001 -v v3.3.0
Error 1749: Downgrading an agent requires the force flag. Use -F to force the downgrade: Agent: v3.7.0 -> v3.3.0
```

### API calls

```
# curl -u foo:bar -k -X PUT "http://127.0.0.1:55000/agents/001/upgrade?pretty&version=v4.0.0"
{
   "error": 1718,
   "message": "Version not available"
}
# curl -u foo:bar -k -X PUT "http://127.0.0.1:55000/agents/001/upgrade?pretty&version=v3.9.0"
{
   "error": 1717,
   "message": "Upgrading an agent to a version higher than the manager requires the force flag. Use -F to force the upgrade"
}
# curl -u foo:bar -k -X PUT "http://127.0.0.1:55000/agents/001/upgrade?pretty&version=v3.0.0"
{
   "error": 1749,
   "message": "Downgrading an agent requires the force flag. Use -F to force the downgrade"
}
```

The `extra_message` part of the message which shows the agent versions is missing by design when upgrading by the API.

**Pull request merged**